File: 2915-count-of-interesting-subarrays/count-of-interesting-subarrays.go
func countInterestingSubarrays(nums []int, modulo int, k int) int64 {
	n := len(nums);
	if k > n {
		return 0;
	}

	prefixCount := make([]int, n+1);
	prefixCount[0] = 1; // base case

	var result int64 = 0;
	prefixModSum := 0;

	for _, num := range nums {
		num %= modulo;
		if num == k {
			prefixModSum++;
		}

		prefixModSum %= modulo;

		targetMod := (prefixModSum - k + modulo) % modulo;

		if targetMod < 0 {
			targetMod += modulo;
		}

		if targetMod < n {
			result += int64(prefixCount[targetMod]);
		}

		prefixCount[prefixModSum]++;
	}

	return result;
}

File: 2915-count-of-interesting-subarrays/count-of-interesting-subarrays.java
class Solution {
    public long countInterestingSubarrays(List<Integer> nums, int modulo, int k) {
        Map<Integer, Long> countMap = new HashMap<>();
        countMap.put(0, 1L); // base case: prefix = 0

        int prefix = 0;
        long result = 0;

        for (int num : nums) {
            // nếu nums[i] % modulo == k thì tăng prefix
            if (num % modulo == k) {
                prefix++;
            }

            // cần tìm: (prefix - k + modulo) % modulo
            int key = (prefix - k + modulo) % modulo;
            result += countMap.getOrDefault(key, 0L);

            // lưu lại prefix % modulo để sau này sử dụng
            int currentMod = prefix % modulo;
            countMap.put(currentMod, countMap.getOrDefault(currentMod, 0L) + 1);
        }

        return result;
    }
}